src/libAtomVM/otp_socket.c

```diff
@@ -0,0 +1,1364 @@
+            if (errno == EINPROGRESS) {
+
+                // TODO make connect non-blocking
+                return UNDEFINED_ATOM;
+
+            } else {
+                AVM_LOGE(TAG, "Unable to connect: res=%i errno=%i", res, errno);
+
+                if (UNLIKELY(memory_ensure_free(ctx, TUPLE_SIZE(2)) != MEMORY_GC_OK)) {
+                    AVM_LOGW(TAG, "Failed to allocate memory: %s:%i.\n", __FILE__, __LINE__);
+                    RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+                }
+
+                term error_tuple = term_alloc_tuple(2, &ctx->heap);
+                term_put_tuple_element(error_tuple, 0, ERROR_ATOM);
+                term_put_tuple_element(error_tuple, 1, globalcontext_make_atom(global, closed_atom));
+                return error_tuple;
+            }
+        } else if (res == 0) {
+            return OK_ATOM;
+        } else {
+            // won't happen according to connect(2)
+            return UNDEFINED_ATOM;
+        }
+    } else {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+}
+
+//
+// Nifs
+//
+
+static const struct Nif socket_open_nif = {
+    .base.type = NIFFunctionType,
+    .nif_ptr = nif_socket_open
+};
+static const struct Nif socket_close_nif = {
+    .base.type = NIFFunctionType,
+    .nif_ptr = nif_socket_close
+};
+static const struct Nif socket_select_stop_nif = {
+    .base.type = NIFFunctionType,
+    .nif_ptr = nif_socket_select_stop
+};
+static const struct Nif socket_setopt_nif = {
+    .base.type = NIFFunctionType,
+    .nif_ptr = nif_socket_setopt
+};
+static const struct Nif socket_bind_nif = {
+    .base.type = NIFFunctionType,
+    .nif_ptr = nif_socket_bind
+};
+static const struct Nif socket_listen_nif = {
+    .base.type = NIFFunctionType,
+    .nif_ptr = nif_socket_listen
+};
+static const struct Nif socket_sockname_nif = {
+    .base.type = NIFFunctionType,
+    .nif_ptr = nif_socket_sockname
+};
+static const struct Nif socket_peername_nif = {
+    .base.type = NIFFunctionType,
+    .nif_ptr = nif_socket_peername
+};
+static const struct Nif socket_select_read_nif = {
+    .base.type = NIFFunctionType,
+    .nif_ptr = nif_socket_select_read
+};
+static const struct Nif socket_accept_nif = {
+    .base.type = NIFFunctionType,
+    .nif_ptr = nif_socket_accept
+};
+static const struct Nif socket_recv_nif = {
+    .base.type = NIFFunctionType,
+    .nif_ptr = nif_socket_recv
+};
+static const struct Nif socket_recvfrom_nif = {
+    .base.type = NIFFunctionType,
+    .nif_ptr = nif_socket_recvfrom
+};
+static const struct Nif socket_send_nif = {
+    .base.type = NIFFunctionType,
+    .nif_ptr = nif_socket_send
+};
+static const struct Nif socket_sendto_nif = {
+    .base.type = NIFFunctionType,
+    .nif_ptr = nif_socket_sendto
+};
+static const struct Nif socket_connect_nif = {
+    .base.type = NIFFunctionType,
+    .nif_ptr = nif_socket_connect
+};
+
+const struct Nif *otp_socket_nif_get_nif(const char *nifname)
+{
+    if (strcmp("socket:open/3", nifname) == 0) {
```

Comment:
Many socket, we may want to first compare with "socket:".

Reply:
Fixed